### PR TITLE
MySql lock hints

### PR DESCRIPTION
**Pull request by igor-tkachev**
#3905

### CODE REVIEW

File: Tests/Base/Remote/TestGrpcLinqService.cs [hunk 1]
```diff
@@ -51,7 +51,7 @@ public DataConnection CreateDataContext(string? configuration)
 			if (_interceptor != null)
 				dc.AddInterceptor(_interceptor);
 
-			if (configuration == ProviderName.MariaDB)
+			if (configuration?.StartsWith(ProviderName.MariaDB) == true)
```
Comment (MaceWindu) [inline Tests/Base/Remote/TestGrpcLinqService.cs:54]: use `configuration.IsAnyOf(TestProvName.AllMariaDB)`

File: Source/LinqToDB/DataProvider/MySql/MySqlHints.cs [hunk 2]
```diff
@@ -595,5 +608,163 @@ public static IMySqlSpecificQueryable<TSource> QueryHint<TSource>(this IMySqlSpe
...
+					var alias = sqlBuilder.BuildSqlID(id);
+
+					stringBuilder.Append(alias);
+				}
+
+				if (sqlQueryExtension.Arguments.TryGetValue("hint2", out var h) && h is SqlValue { Value: string value })
+				{
+//					if (value != SubQuery.SkipLocked
+//						|| sqlBuilder.MappingSchema.ConfigurationList.Contains(ProviderName.PostgreSQL95)
+//						|| sqlBuilder.MappingSchema.ConfigurationList.Contains(ProviderName.PostgreSQL15))
```
Comment (MaceWindu) [inline Source/LinqToDB/DataProvider/MySql/MySqlHints.cs:643]: remove

File: Source/LinqToDB/ProviderName.cs [hunk 3]
```diff
@@ -115,6 +115,11 @@ public static class ProviderName
 		/// </summary>
 		public const string MySqlConnector = "MySqlConnector";
 		/// <summary>
+		/// MySql provider.
+		/// Used as an additional configuration name for MySql mapping schema <see cref="DataProvider.MySql.MySqlMappingSchema"/>.
```
Comment (MaceWindu) [inline Source/LinqToDB/ProviderName.cs:119]: should reference MariaDB, not MySql

File: Tests/Base/Remote/ServerContainer/GrpcServerContainer.cs [hunk 4]
```diff
@@ -62,14 +63,18 @@ public GrpcServerContainer()
 			Debug.WriteLine(((IDataContext) dx).ConfigurationID, "Provider ");
 
 			if (ms != null)
-				dx.MappingSchema = dx.MappingSchema == null ? ms : MappingSchema.CombineSchemas(ms, dx.MappingSchema);
+				dx.MappingSchema = MappingSchema.CombineSchemas(ms, dx.MappingSchema);
+
+			if (configuration.StartsWith(TestProvName.MariaDB))
```
Comment (MaceWindu) [inline Tests/Base/Remote/ServerContainer/GrpcServerContainer.cs:68]: 1. if we add MariaDB to `ProviderName` we should remove it from `TestProvName`
2. use `configuration.IsAnyOf(TestProvName.AllMariaDB)` to check for mariadb test database

File: Tests/Base/Remote/ServerContainer/TestLinqService.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace Tests.Remote.ServerContainer
+{
+	public class TestLinqService : LinqService
+	{
+		public override DataConnection CreateDataContext(string? configuration)
+		{
+			var dc = base.CreateDataContext(configuration);
+
+			if (configuration?.StartsWith(ProviderName.MariaDB) == true)
```
Comment (MaceWindu) [inline Tests/Base/Remote/ServerContainer/TestLinqService.cs:15]: use `configuration.IsAnyOf(TestProvName.AllMariaDB)` to check for mariadb test database

File: Tests/Base/TestBase.cs [hunk 6]
```diff
@@ -476,8 +476,8 @@ protected TestDataConnection GetDataConnection(string configuration, Func<DataOp
 
 				options = options.UseMappingSchema(ms == null ? _sequentialAccessSchema : MappingSchema.CombineSchemas(ms, _sequentialAccessSchema));
 			}
-			//else if (configuration == TestProvName.SqlServer2019FastExpressionCompiler)
-			//	res.AddMappingSchema(_fecSchema);
+			else if (configuration.StartsWith(TestProvName.MariaDB))
```
Comment (MaceWindu) [inline Tests/Base/TestBase.cs:479]: use `configuration.IsAnyOf(TestProvName.AllMariaDB)` to check for mariadb test database

**Review by MaceWindu — CHANGES_REQUESTED** [review]

File: Tests/Base/Remote/TestGrpcLinqService.cs (hunk 1 above)
Comment (igor-tkachev) [reply Tests/Base/Remote/TestGrpcLinqService.cs]: fixed

File: Source/LinqToDB/ProviderName.cs (hunk 3 above)
Comment (igor-tkachev) [reply Source/LinqToDB/ProviderName.cs:119]: Do we have MariaDB provider?

File: Source/LinqToDB/DataProvider/MySql/MySqlHints.cs (hunk 2 above)
Comment (igor-tkachev) [reply Source/LinqToDB/DataProvider/MySql/MySqlHints.cs:643]: done

File: Tests/Base/Remote/ServerContainer/GrpcServerContainer.cs (hunk 4 above)
Comment (igor-tkachev) [reply Tests/Base/Remote/ServerContainer/GrpcServerContainer.cs]: fixed

File: Tests/Base/Remote/ServerContainer/TestLinqService.cs (hunk 5 above)
Comment (igor-tkachev) [reply Tests/Base/Remote/ServerContainer/TestLinqService.cs]: fixed

File: Tests/Base/TestBase.cs (hunk 6 above)
Comment (igor-tkachev) [reply Tests/Base/TestBase.cs]: fixed

**Review by MaceWindu — APPROVED** [review]

**Review by sdanyliv — APPROVED** [review]

File: Source/LinqToDB/DataProvider/MySql/MySqlHints.cs [hunk 7]
```diff
@@ -595,5 +608,172 @@ public static IMySqlSpecificQueryable<TSource> QueryHint<TSource>(this IMySqlSpe
...
+
+		sealed class SubQueryTableHintExtensionBuilder : ISqlQueryExtensionBuilder
+		{
+			void ISqlQueryExtensionBuilder.Build(ISqlBuilder sqlBuilder, StringBuilder stringBuilder, SqlQueryExtension sqlQueryExtension)
+			{
+				var hint    = (string)((SqlValue)sqlQueryExtension.Arguments["hint"]).Value!;
+				var idCount = (int)   ((SqlValue)sqlQueryExtension.Arguments["tableIDs.Count"]).Value!;
+
+				if ((hint is SubQuery.ForShare || idCount > 0) && sqlBuilder.MappingSchema.ConfigurationList.Contains(ProviderName.MariaDB))
+					stringBuilder.Append("-- ");
```
Comment (MaceWindu) [inline Source/LinqToDB/DataProvider/MySql/MySqlHints.cs:622]: What does it mean? Why we generate hint as comment?

Comment (igor-tkachev) [reply Source/LinqToDB/DataProvider/MySql/MySqlHints.cs:622]: It means that this sql dialect does not support this feature.

**Pull request merged**